### DISCUSSION TITLE
Add support to Iotty OSWFR

### DISCUSSION THREAD

**Discussion by DzzD** [arendst/Tasmota#17206] [Q&A]
Hello, 

First I have to say that I am really noob to tasmota, but I have strong knowledge on programming and high on electronic, and I have read a lot about tasmota, so I think I begin to understand how it works.

Here is my question/suggestion, I recently bought a Lotty OSFWR, and despite its price (from 35€ to 80€...) I think it give a lot of advantage :
- nice design
- recessed socket
- button to turn on/off

I openend it and it seems to run on an ESP32, my question is it already supported by Tasmota ? (I did not found it on the list)
![image](https://user-images.githubusercontent.com/12198339/204280322-c1ad9df5-dd51-4aa4-9467-f8042fb418c0.png)

If no what is the process to do that ? I suppose at first I need to find a way to to flash the firmware, and identify pin I/O ?

Another question, is there other recessed socket with an on/off button already available that work easily with Tasmota (I did not find any) ?

**Comment by barbudor** [comment]
Hi @DzzD 
I can't neither find any Lotti devices in the template repository 

So it's all up to you to find out which function is connected to each GPIO
I trust that with your skills in electronics you should have no problem tracing the tracks
Here is also a little guide meant for beginniers: https://tasmota.github.io/docs/Configuration-Procedure-for-New-Devices/

> Another question, is there other recessed socket with an on/off button already available that work easily with Tasmota (I did not find any) ?

Not sure what you means with "recessed socket". is that a socket that goes in the wall ?
May be you can add a photo ?
In most of the case the idea would be to use a standard one with mini device behind like Sonoff Mini, Shelly, ....

I've seen such devices on AliExpress but never ordered one to see if it can be flashed with Tasmota. This irsk is high to have an non-ESP on board.

One clue could be to have a look or ask to manufacturers already selling pre-flashed device is they have such device : https://templates.blakadder.com/preflashed.html

**Reply by DzzD** [reply]
_**So it's all up to you to find out which function is connected to each GPIO
I trust that with your skills in electronics you should have no problem tracing the tracks
Here is also a little guide meant for beginniers: https://tasmota.github.io/docs/Configuration-Procedure-for-New-Devices/**_

No option/menu in tasmota to make a generic diagnostic, it would be really handy. I mean a menu where you can read/write any GPIO, this would make new devices easier to configure and would only requiere to flash them once with tasmota, no need of other electronic DIY, just a simple tasmota flash.

Oups, seems your link explain that, sorry

_**Not sure what you means with "recessed socket". is that a socket that goes in the wall ?
May be you can add a photo ?
In most of the case the idea would be to use a standard one with mini device behind like Sonoff Mini, Shelly, ....**_

Yes I mean "in the wall" by "recessed" sorry I dont know the right word in english (can you tell me wich is the right word plz ?), problem with existing module like sonoff Mini that goes inside the wall behind the socket plate is that they dont offer a button to manually turn on/off the socket, (it exist on a lot of external socket not inside the wall and it is very handy, but I would prefer an into the wall socket)
![image](https://user-images.githubusercontent.com/12198339/204300807-1134efc9-5b1d-48cb-b677-4aa683514435.png)

**Comment by DzzD** [comment]
Well, it begin very well, they put some circuit test points, and even better they put debug pinout to solder plug, which is not the case of all manufacturer !

Last but not least, this thing have been made to be disassembled !

![LottyOSWFR-1](https://user-images.githubusercontent.com/12198339/204324817-d0744c6c-9f75-41a2-b711-539678c31331.png)
![LottyOSWFR-2](https://user-images.githubusercontent.com/12198339/204324842-b2607400-15e1-4cab-9152-0c0571b732c4.png)
![LottyOSWFR-3](https://user-images.githubusercontent.com/12198339/204324859-6487698d-99b6-46cc-b90c-daad64dad72f.png)
![LottyOSWFR-4](https://user-images.githubusercontent.com/12198339/204324870-59cea8d5-970b-491b-9c60-552d6ac54d4b.png)

I am studing the pinout now, I will post progress & explanation in this discussion

**Reply by DzzD** [reply]
Ok, I think I found the pins to use for serial programming, still one I am not sure.

![ESP32-VROOM-32D-PINOUT-ZOOM](https://user-images.githubusercontent.com/12198339/204349945-a0080901-03d3-4d12-85d1-6a55d5ed51fe.png)

![LottyOSWFR-PCBFront](https://user-images.githubusercontent.com/12198339/204350083-c32a4a5e-7de0-492f-b331-d197346fce69.png)

![LottyOSWFR-PCBBack](https://user-images.githubusercontent.com/12198339/204350100-ae5417d0-c3a3-4429-894a-afa6e6d70efd.png)

![LottyOSWFR-PCBBackFront](https://user-images.githubusercontent.com/12198339/204350116-b2aab4ac-6162-466c-b0ce-46776903b3b6.png)


I did not powerup anything for now, what could be the pull-upped pin for ? reset ? I'll try tomorrow to flash it

NB: 
- the photo are very zoomed, the tx & rx probe point are less than a milimeter...)

- I have joined the PSD images source file, if someone interrested
[LottyOSWFR.zip](https://github.com/arendst/Tasmota/files/10106459/LottyOSWFR.zip)

ERRATUM : the VCC at the upper-right corner is a GND... mistyping

**Reply by sfromis** [reply]
For flashing, you normally need gnd, vcc/3.3v, tx, rx and gpio0. First 4 connected to the USB/serial adapter, and gpio0 connected to ground while applying power. If you are lucky, the ??? might be gpio0? Otherwise, easy to find on the ESP32 module.

Make sure to use a USB/serial adapter with at least 150 mA of 3.3V power (like the recommended one), many of those adapters fail in this regard.

**Reply by DzzD** [reply]
Arff I got only an FT232 it won't be able to draw that much current, I'll need an external power supply, thanks for the tips, as for the unkwon pin I think you are right as it is pulled-up and also even if it is not a strong proof the ohmmeter give 0Ohm between the ESP32 chip GPIO0 and the ??? pin

**Reply by DzzD** [reply]
Well thinking/looking on it a bit more, I think that I have everything now, there are two maintenance plug: 
The 5 horizontal holes is a plug to progam it : with GND/GPIO0/RX/TX/VCC
The vertical holes is a plug for debugging (from top to down) : GND/GPIO22/GPIO23/VCC

Tomorrow, I will reassamble everything, power the socket just to double check a last time if I am right for VCC & GND, then I ll disasamble and try to flash it from USB using and additional external 3.3v power supply.

If I am right, I guess the only GPIO used should probably be the 22 & 23, but for now it is only pure theory :) :)

**Reply by sfromis** [reply]
> even if it is not a strong proof the ohmmeter give 0Ohm between the ESP32 chip GPIO0 and the ??? pin

That sounds like reasonably good proof, I'd go for it

**Reply by sfromis** [reply]
I see several traces from the ESP32 to that 8-pin connector right above C2-DBG. Likely interesting. Of course also what happens on that flat ribbon cable.

I'm a bit concerned about what that small 16-pin U2 chip in the middle is doing. SB10 F8G seems to be a small MCU. Looks like it is connected to TX/RX. That could be "additional complications".

**Reply by DzzD** [reply]
> I'm a bit concerned about what that small 16-pin U2
U2 ? That's probably all just about music :) 

You are right tx/rx are in fact connected to this little U2 chip, the question is why they needed to put this chip ? afterall is is only a switch...  it may be related to the touch plate that enable shut on/off of the socket or another of its functionalities ?

I did not see it when I bought it but the switch offer more functionalities that I was thinking at first like : programmable retro-ligth / touch screen (on/off) / power consumption. But nothing that the ESP32 cant handle ?

**EDIT :** 
I was in too much of a hurry... and too much concentrated on a single part of the ci... the conclusion is rather than GPIO 22 & 23 are probably not used or have a specific usage, that would explain why there are on the debug plug. Also a cupple of other GPIO (4,16,17,5) seems to go straight to the black plug....

Here is a similar question, this chips is around 2,5mmx2,5mm, but references look but not same as this discussion : 
https://electronics.stackexchange.com/questions/443364/itead-sonoff-rf-wifi-433mhz-switch-identifying-an-unknown-chip-on-the-433mhz



I got no idea of what the ribbon cable is intended for but probably exchange data with the plug

**Reply by sfromis** [reply]
Sometimes watching pin activity with factory firmware (do remember to first make a backup if you flash Tasmota) can be a "learning experience" about how the device works.

**Reply by sfromis** [reply]
The chip mentioned in your edit is BB10, not SB10, but still a small MCU. As such a MCU is general purpose, you should anyway not focus too much of some particular use case for such a chip.

Here's the specs for [SB10 F8G](https://www.silabs.com/mcu/8-bit-microcontrollers/efm8-sleepy-bee/device.efm8sb10f8g-csp16), or full name EFM8SB10F8G (also exists in a variant with more pins along the edges). The application notes do have links about "capacitive sensing", which makes it plausible that the chip is used in relation to touch.

**Reply by DzzD** [reply]
Not sure but 

> The chip mentioned in your edit is BB10, not SB10, but still a small MCU. As such a MCU is general purpose, you should anyway not focus too much of some particular use case for such a chip.
> 
> Here's the specs for [SB10 F8G](https://www.silabs.com/mcu/8-bit-microcontrollers/efm8-sleepy-bee/device.efm8sb10f8g-csp16), or full name EFM8SB10F8G (also exists in a variant with more pins along the edges). The application notes do have links about "capacitive sensing", which makes it plausible that the chip is used in relation to touch.

I have searched and I was unable to find the exact same reference, but as you said and you're right, It is an MCU general purpose and could be used for many thing, details are not that important, what we know is that it is connected to rx/tx, it is ultra low consumption, and may have touch capabilities... the investigation is progressing :)

**Reply by DzzD** [reply]
> Sometimes watching pin activity with factory firmware (do remember to first make a backup if you flash Tasmota) can be a "learning experience" about how the device works.

Well, this is far out of my possibilities given the tools available to me, the chip itself is less than 2,5mm with 4 pin on each side, not to mention that I have only some basic tools, nothing fancy, just enought to play. (side note, I love this french channel https://www.youtube.com/channel/UCH6ppHEvV3_WIXEwmhv9HEg, this guy is able to xray any IC and give a complete retro enginering of nearly anything)

**Reply by sfromis** [reply]
When you have the tx/rx pins, you have a good and easily accessible point for looking at activity.

**Reply by DzzD** [reply]
> When you have the tx/rx pins, you have a good and easily accessible point for looking at activity.

Well, once again you are absolutlty right for those two pins, not having an oscilloscope, I can probably log them using two input of another ESP using the RX/TX probes points, with a simple program sending those to the PC USB, but I was not at first thinking of so much time/work even if it is very interresting and I would like to spent much more time on this hobby.

So what I could do that may be faster but a bit more risky is maybe backup the firmware, try to flash it and see if nothing is broken and if basic functions can works :) 

If something goes wrong I will try to go back to initial firmware for trying to understand what ESP and MCU tell to each other ? 

Does it sound a good plan for you ?

Thanks for all your advices and time

**Reply by DzzD** [reply]
I make a bet, so I can lose, but if the mcu only send data to ESP it should be ok... hopes...

**Reply by sfromis** [reply]
One possible scenario is the touch not working without handling data from the SB10 F8G, but relay still available to Tasmota. Still, there are  quite a few other pins of possible interest, outside of the simple relay.

**Reply by DzzD** [reply]
I made a little more disassembly and I know what the ribbon cable is for, it goes straight to the front plate, so it is used for light and power button.

I dont understand why they would need something more than the ESP for lightning, but maybe for the sensitive touch switch it requiere the MCU. I ll definitly study what happen with the stock firmware before going further.

The black plug is connected to the power block, so I assume it is responsible of switching on/off as well as reading if it is on or off and also to measure the consumption, it is probably all managed by the ESP.

**Reply by DzzD** [reply]
I have the feeling that tx & rx are used as regular GPIO for turning light on & off and receiving on & off plate switch, going through the MCU for an unknown reason., just a guess.

**Reply by sfromis** [reply]
My uninformed guess would be that tx/rx are used for communication between ESP32 and the other MCU, plausibly via serial communication, which is *very* common for communication in a dual-MCU setup.

**Reply by DzzD** [reply]
//side note start
This Tyny MCU also got their own DIY progamming board, but not an option for me, way too small, too much time, no interrest, I would prefere if requiered to bypass it, I was just thinking that was interresting to know. 

https://www.silabs.com/development-tools/mcu/8-bit/efm8bb1lck-starter-kit?tab=overview

Also the ref can be decoded easily : : 
![20221201_004947](https://user-images.githubusercontent.com/12198339/204933153-0eb304d0-522f-466e-9737-db32b36bfaf8.jpg)

And the documentation is also available for this family of MCU, by references decoding it give the more precise capabilities/featured

Adding the datashhetne fo family gave néerlandais everyrhing to know, I ll Nevers try ass it is to small but if anyone do a try & succession for the fun, I would be really interested

https://www.silabs.com/documents/public/data-sheets/efm8sb1-datasheet.pdf


//side not end

**Comment by DzzD** [comment]
short word to give some news, I did not have enough time to finish this WE, will do next WE
